### PR TITLE
120 fix redirect with variable

### LINKED ISSUE #120: Fix redirect with variable
## Example

``` bash
minishell$ export VAR='out.txt'
minishell$ echo Hello > $VAR
minishell$ ls '$VAR'
$VAR
```

### PR DESCRIPTION
**Pull request by rask24**
fix #120

- **Implement set_file_or_delim**
- **Fix handle_redirect**
- **Rename redirects to redirect**
- **Add e2e comments**

## Summary by CodeRabbit

- **New Features**
	- Introduced a new function for setting file or delimiter in redirect handling.
	- Enhanced the abstract syntax tree (AST) manipulation capabilities.
	- Added new test functions for heredoc functionality with variable and wildcard delimiters.
	- Expanded test coverage for output redirection scenarios.

- **Bug Fixes**
	- Streamlined redirect handling to improve clarity and reduce complexity.
	- Improved command execution logic to prevent potential errors.

- **Tests**
	- Added new test cases for the `set_file_or_delim` function.
	- Added tests for output redirection errors.
	- Added tests for handling heredocs with various delimiters.
	- Commented out inactive test functions related to redirection with environment variables.

### CODE REVIEW

File: src/exec/handle_redirect.c [hunk 1]
```diff
@@ -39,13 +39,13 @@ static int	open_redirect_file(t_redirect_info *info, t_ctx *ctx)
 	return (fd);
 }
 
-static const char	*expand_filepath(t_list *redirects, t_ctx *ctx)
+static const char	*expand_filepath(t_list *redirect, t_ctx *ctx)
```
Comment (alex216) [inline src/exec/handle_redirect.c:42]:
```suggestion
static const char	*expand_filepath(const char *filepath, t_ctx *ctx)
```

File: src/exec/handle_redirect.c [hunk 2]
```diff
@@ -39,13 +39,13 @@ static int	open_redirect_file(t_redirect_info *info, t_ctx *ctx)
...
 {
 	t_list		*original;
 	t_list		*files;
 	const char	*ret;
 
-	original = ft_xlstnew(ft_xstrdup(get_redirect_file_or_delim(redirects)));
+	original = ft_xlstnew(ft_xstrdup(get_redirect_file_or_delim(redirect)));
```
Comment (alex216) [inline src/exec/handle_redirect.c:48]:
```suggestion
	original = ft_xlstnew(ft_xstrdup(filepath));
```

File: src/exec/handle_redirect.c [hunk 3]
```diff
@@ -70,28 +70,28 @@ static void	redirect_std_fd(t_redirect_type type, int fd)
...
 		return (false);
 	}
-	fd = open_redirect_file(redirects->content, ctx);
+	fd = open_redirect_file(redirect->content, ctx);
 	if (fd == -1)
 	{
-		if (get_redirect_type(redirects) != REDIRECT_HEREDOC)
-			print_error(expanded_filepath, strerror(errno));
+		if (get_redirect_type(redirect) != REDIRECT_HEREDOC)
+			print_error(get_redirect_file_or_delim(redirect), strerror(errno));
```
Comment (alex216) [inline src/exec/handle_redirect.c:90]:
```suggestion
			print_error(get_redirect_file_or_delim(redirect), strerror(errno));
free((char *)filepath);
```

**Comment by alex216** [comment]
error case ?
```bash
minishel$ << EOF
> text
> EOF
[1]    14498 segmentation fault  ./minishell
```

```bash
bash-5.2$ << EOF
> text
> EOF
bash-5.2$
```

**Comment by alex216** [comment]
@rask24 please check the comments.

File: src/exec/handle_redirect.c (hunk 1 above)
Comment (rask24) [reply src/exec/handle_redirect.c:42]: I've updated the function. Please review it!

File: src/exec/handle_redirect.c (hunk 2 above)
Comment (rask24) [reply src/exec/handle_redirect.c:50]: same above

File: src/exec/execute_command.c [hunk 4]
```diff
@@ -95,7 +95,7 @@ int	execute_command(t_ast *node, t_ctx *ctx, t_pipeline_conf *conf)
 	expanded_cmd_args = expand(node->cmd_args, ctx);
 	ft_lstclear(&node->cmd_args, free);
 	node->cmd_args = expanded_cmd_args;
-	if (is_builtin(get_cmd_arg(node->cmd_args)))
+	if (node->cmd_args == NULL || is_builtin(get_cmd_arg(node->cmd_args)))
```
Comment (alex216) [inline src/exec/execute_command.c:98]: wrong exit_status..

```bash
minishell$ << a
> a
minishell$ echo $?
1
```
```bash
bash-5.2$ << a
> a
bash-5.2$ echo $?
0
```

File: src/exec/handle_redirect.c [hunk 5]
```diff
@@ -70,28 +72,29 @@ static void	redirect_std_fd(t_redirect_type type, int fd)
 	}
 }
 
-bool	handle_redirect(t_list *redirects, t_ctx *ctx)
+bool	handle_redirect(t_list *redirect, t_ctx *ctx)
```
Comment (alex216) [inline src/exec/handle_redirect.c:75]: suggestion
```c
bool	handle_redirect(t_list *redirect, t_ctx *ctx)
{
	int			fd;
	const char	*expanded_filepath;

	expanded_filepath = expand_filepath(redirect, ctx);
	if (expanded_filepath == NULL)
	{
		print_error(get_redirect_file_or_delim(redirect), "ambiguous redirect");
		set_file_or_delim(redirect, NULL);
		return (false);
	}
	set_file_or_delim(redirect, expanded_filepath);
	fd = open_redirect_file(redirect->content, ctx);
	if (fd == -1)
	{
		if (get_redirect_type(redirect) != REDIRECT_HEREDOC)
			print_error(get_redirect_file_or_delim(redirect), strerror(errno));
		return (false);
	}
	redirect_std_fd(get_redirect_type(redirect), fd);
	close(fd);
	return (true);
}
```

Comment (rask24) [reply src/exec/handle_redirect.c:75]: It is better than current one.
But it is need to add the below line.
free((char *)get_redirect_file_or_delim(redirect));

File: src/exec/execute_command.c [hunk 6]
```diff
@@ -58,6 +58,10 @@ static void	restore_std_io(int *std_fds)
 	close(std_fds[2]);
 }
 
+/*
+** argv is NULL only if no arguments are passed to the command.
+** In this case, this function returns EXIT_SUCCESS.
+*/
 static int	execute_builtin_command(t_ast *node, t_ctx *ctx)
```
Comment (alex216) [inline src/exec/execute_command.c:65]: radical suggestion to use return early pattern.

```c
static int	execute_builtin_command(t_ast *node, t_ctx *ctx)
{
	const char	*cmd_name;
	char		**argv;
	int			status;

	status = 0;
	cmd_name = get_cmd_arg(node->cmd_args);
	argv = convert_cmd_args_to_array(node->cmd_args);
	if (ft_strcmp(cmd_name, "cd") == 0)
		status = builtins_cd(argv, ctx);
	else if (ft_strcmp(cmd_name, "pwd") == 0)
		status = builtins_pwd(argv, ctx);
	else if (ft_strcmp(cmd_name, "echo") == 0)
		status = builtins_echo(argv, ctx);
	else if (ft_strcmp(cmd_name, "env") == 0)
		status = builtins_env(argv, ctx);
	else if (ft_strcmp(cmd_name, "export") == 0)
		status = builtins_export(argv, ctx);
	else if (ft_strcmp(cmd_name, "unset") == 0)
		status = builtins_unset(argv, ctx);
	else if (ft_strcmp(cmd_name, "exit") == 0)
		status = builtins_exit(argv, ctx);
	ft_free_strs(argv);
	return (status);
}

int	execute_command(t_ast *node, t_ctx *ctx, t_pipeline_conf *conf)
{
	int		std_fds[3];
	t_list	*expanded_cmd_args;

	if (node->cmd_args == NULL)
	{
		ctx->exit_status = EXIT_SUCCESS;
		return (EXIT_SUCCESS);
	}
	expanded_cmd_args = expand(node->cmd_args, ctx);
	ft_lstclear(&node->cmd_args, free);
	node->cmd_args = expanded_cmd_args;
	if (is_builtin(get_cmd_arg(node->cmd_args)))
	{
		save_std_io(std_fds);
		if (handle_io(conf, node->redirects, ctx))
			ctx->exit_status = execute_builtin_command(node, ctx);
		else
			ctx->exit_status = EXIT_FAILURE;
		restore_std_io(std_fds);
		return (EXIT_SUCCESS);
	}
	else
		return (spawn_command(node, ctx, conf));
}
```

File: tests/e2e/test_heredoc.py [hunk 7]
```diff
@@ -42,6 +42,20 @@ def test_heredoc_expand(shell_session):
     assert result == "Hello!!"
 
 
+def test_heredoc_expnad_quoted(shell_session):
```
Comment (alex216) [inline tests/e2e/test_heredoc.py:45]:
```suggestion
def test_heredoc_expand_quoted(shell_session):
```

File: src/exec/execute_command.c (hunk 6 above)
Comment (rask24) [reply src/exec/execute_command.c:65]: So `handle_io` must always be applied.

**Pull request merged**